### PR TITLE
Ticket3585: Autoscales log plot for a short time when plot opened

### PR DESCRIPTION
**Pull request by adamb56789**
### Description of work

Set auto-scale to true by default so it scales the plot when it is first opened, then added an event listener to turn auto-scale back off after 10 data points are added.

### Ticket

https://github.com/ISISComputingGroup/IBEX/issues/3585

### Acceptance criteria

- [x] Auto-scale works a single time when a new graph is opened.

### Documentation

https://github.com/ISISComputingGroup/ibex_user_manual/wiki/Plot-a-Block-Graph
Made some changes to clarify some of the things mentioned in the ticket.

---

#### Code Review

- [ ] Is the code of an acceptable quality?
- [ ] Do the changes function as described and is it robust?
- [ ] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev)?

### Final Steps
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

**Comment by DominicOram** [comment]
This works as described but has a couple of issues. I occasionally now get the following exception:
```
Exception in thread "ForkJoinPool-1-worker-0" org.eclipse.swt.SWTException: Invalid thread access
	at org.eclipse.swt.SWT.error(SWT.java:4552)
	at org.eclipse.swt.SWT.error(SWT.java:4467)
	at org.eclipse.swt.SWT.error(SWT.java:4438)
	at org.eclipse.swt.widgets.Widget.error(Widget.java:448)
	at org.eclipse.swt.widgets.Widget.checkWidget(Widget.java:366)
	at org.eclipse.swt.widgets.Table.deselectAll(Table.java:1944)
	at org.eclipse.jface.viewers.TableViewer.doDeselectAll(TableViewer.java:300)
	at org.eclipse.jface.viewers.AbstractTableViewer.setSelectionToWidget(AbstractTableViewer.java:830)
	at org.eclipse.jface.viewers.StructuredViewer.setSelectionToWidget(StructuredViewer.java:1760)
	at org.eclipse.jface.viewers.StructuredViewer.preservingSelection(StructuredViewer.java:1453)
	at org.eclipse.jface.viewers.StructuredViewer.preservingSelection(StructuredViewer.java:1407)
	at org.eclipse.jface.viewers.StructuredViewer.refresh(StructuredViewer.java:1507)
	at org.eclipse.jface.viewers.ColumnViewer.refresh(ColumnViewer.java:523)
	at org.eclipse.jface.viewers.StructuredViewer.refresh(StructuredViewer.java:1466)
	at org.csstudio.trends.databrowser2.propsheet.TraceTableHandler$1.changedAxis(TraceTableHandler.java:78)
	at org.csstudio.trends.databrowser2.model.Model.fireAxisChangedEvent(Model.java:936)
	at org.csstudio.trends.databrowser2.model.AxisConfig.fireAxisChangeEvent(AxisConfig.java:305)
	at org.csstudio.trends.databrowser2.model.AxisConfig.setAutoScale(AxisConfig.java:285)
	at uk.ac.stfc.isis.ibex.ui.logplotter.LogPlotterHistoryPresenter$1.selectedSamplesChanged(LogPlotterHistoryPresenter.java:128)
	at org.csstudio.trends.databrowser2.model.Model.fireSelectedSamplesChanged(Model.java:975)
	at org.csstudio.trends.databrowser2.ui.Controller$3.selectedSamplesChanged(Controller.java:386)
	at org.csstudio.trends.databrowser2.ui.ModelBasedPlot$2.lambda$3(ModelBasedPlot.java:135)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.csstudio.trends.databrowser2.ui.ModelBasedPlot$2.changedCursors(ModelBasedPlot.java:135)
	at org.csstudio.swt.rtplot.internal.Plot.fireCursorsChanged(Plot.java:1296)
	at org.csstudio.swt.rtplot.internal.Plot.updateCursors(Plot.java:1119)
	at org.csstudio.swt.rtplot.internal.PlotProcessor.lambda$1(PlotProcessor.java:297)
	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```
I can somewhat reliably reproduce it by adding two traces to the graph and getting them to share an axis then add another 4~5 traces to the graph.

It would probably be useful to add a comment to the code to say what the listener is doing and why.

This doesn't address the 2nd issue on the original ticket about linking the axes. Can you either address this (maybe by adding the option in the block context menu for which axis to plot on) or pull it into a separate ticket.

**Comment by adamb56789** [comment]
I added a comment about the listener and a new option in the menu to add to an existing axis (and updated the [wiki](https://github.com/ISISComputingGroup/ibex_user_manual/wiki/Plot-a-Block-Graph)).
I encountered the same exception as above during testing, though it did not take that many traces for it to occur. One time I created a new plot and got the error immediately.

**Comment by adamb56789** [comment]
I've followed the internet's instructions and put the offending line in an asyncExec.
```java
Display.getDefault().asyncExec(new Runnable() {
     public void run() {
        axis.setAutoScale(false);
    }
});
```
I made several attempts to reproduce the bug after the change, and I didn't get it to happen.

File: base/uk.ac.stfc.isis.ibex.ui.blocks.presentation/src/uk/ac/stfc/isis/ibex/ui/blocks/presentation/AxisList.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+package uk.ac.stfc.isis.ibex.ui.blocks.presentation;
+
+import java.util.stream.Stream;
+
+/**
+ * A class which holds the name of a plot and its axes.
+ */
+public class AxisList {
+    private String plotName;
+    private Stream<String> axisNames;
```
[the remark this answers is not in the document]
Comment (adamb56789) [reply base/uk.ac.stfc.isis.ibex.ui.blocks.presentation/src/uk/ac/stfc/isis/ibex/ui/blocks/presentation/AxisList.java]: I've changed it.

**Comment by DominicOram** [comment]
This works well and the exception I was getting previously is now gone. However, there are a few bits of functionality I would change:
* The context menu is getting very long now can we instead have submenus so that I can click on `Display Block History` -> `Add to XYZ plot` -> `Add to new axis`
* The colour of the axes is now getting confusing. Maybe if you add a second trace to an axis it's colour changes to a neutral black? This is a nitpick though.

I also think the code could be a little tidier:
* The LogPlotterHistoryPresenter class now has quite a bit of duplicated code, specifically between `addPVToEditor` and `addPVToAxis` and between `addToDisplay` and `addToAxis`. Can you try and refactor this to reduce the duplication? Maybe have a look at [Optional](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html), which might help.
* I think you could replace the `AxisList` class with a HashMap.
* Rather than have the two methods `getDataBrowserTitles` and `getAxisTitles` you can probably just have one as all the information for `getDataBrowserTitles` is in `getAxisTitles`
* I've added a couple of minor points in the code too.

File: base/uk.ac.stfc.isis.ibex.ui.logplotter/src/uk/ac/stfc/isis/ibex/ui/logplotter/LogPlotterHistoryPresenter.java [hunk 2]
```diff
@@ -113,13 +126,79 @@ private void addPVToEditor(String pvAddress, final String displayName, DataBrows
...
+        // Add received items
+        final double period = Preferences.getScanPeriod();
+        try {
+            // Create trace
+            final PVItem item = new PVItem(pvAddress, period);
+            item.setDisplayName(displayName);
+            item.useDefaultArchiveDataSources();
+
+            // Extract the axis to add to from the list of axes
+            List<AxisConfig> axes = StreamSupport.stream(model.getAxes().spliterator(), false).filter(a -> a.getName() == axisName).collect(Collectors.toList());
```
Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.ui.logplotter/src/uk/ac/stfc/isis/ibex/ui/logplotter/LogPlotterHistoryPresenter.java]: You only want one axis, take a look at the `findFirst` method

File: base/uk.ac.stfc.isis.ibex.ui.logplotter/src/uk/ac/stfc/isis/ibex/ui/logplotter/LogPlotterHistoryPresenter.java [hunk 3]
```diff
@@ -113,13 +126,79 @@ private void addPVToEditor(String pvAddress, final String displayName, DataBrows
...
+
+			    // This listener is triggered every time that a data point is added to the graph,
+                // either from archive data or from new data. After 10 pieces of data are received,
+			    // it turns off autoscale.
+	            @Override
+	            public void selectedSamplesChanged() {
+	                if (dataCount < DATA_COUNT_LIMIT) {
+	                    dataCount++;
+	                } else if (autoscale) {
+	                    Display.getDefault().asyncExec(new Runnable() {
```
Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.ui.logplotter/src/uk/ac/stfc/isis/ibex/ui/logplotter/LogPlotterHistoryPresenter.java]: There's some slightly leaner syntax you can use here: `Display.getDefault().asyncExec(() -> axis.setAutoScale(false));`

**Comment by adamb56789** [comment]
I've made all of the suggested changes.

**Pull request merged**